### PR TITLE
Add an optional maximum length to Text field types

### LINKED ISSUE #2488: [EN] Add length limitation to Text FieldType
### Description

This should be a quick PR the next chance I have but wanted to get it here for any discussion.  It would be nice to be able to set a max length on the Text FieldType.

### Steps to Reproduce

1. Create a Person attribute of Text field type
2. Add a core category and Save
3. Go to the attribute and add a wall of Lorem Ipsum text
4. Groan with agony as you didn't want someone's life story stored in that attribute.

**Expected behavior:** 

Be able to set a max length for a text field, as this functionality is supported on all TextBox controls.

**Actual behavior:** 

Not able to set a max length for a text field.  Confirmed functionality not available in pre-alpha.

### PR DESCRIPTION
**Pull request by dcs619**
## Contributor Agreement
_Have you filled out and sent your [Spark Contributor Agreement](http://www.rockrms.com/Content/RockExternal/Misc/Contributor%20Agreement.pdf) to secretary [at] sparkdevnetwork.org?_

✅

## Context
_What is the problem you encountered that lead to you creating this pull request?_

Fix #2488.

## Goal
_What will this pull request achieve and how will this fix the problem?_

Adds support for the pre-existing Textbox property of Max Length to the Rock Text fieldtype.

## Screenshots
_Provide us some screenshots if your pull request either alters existing UI or provides new UI. Arrows and labels are helpful._

screenshot 2017-10-23 23 00 34

screenshot 2017-10-23 22 58 16

## Documentation
_If your change effects the UI or needs to be documented in one of the existing [user guides](http://www.rockrms.com/Learn/Documentation), please provide the brief write-up here:_

An attribute of type Text can now be set with a maximum input length.  Keep in mind that the user isn't prompted when they reach the end of the field; they just won't be able to type any more.

## Migrations
Should your pull request require a migration, please exclude the migration from the Rock.Migration project, but submit it in your pull request. Please add a note to your pull request that provides a heads up that a migration file is present.

N/A

### CODE REVIEW

**Comment by edmistj** [comment]
@dcs619 could you make the configuration be 'leave blank for unlimited' vs setting it to 0.

**Comment by dcs619** [comment]
@edmistj the code already ignores Max Length if it's blank or 0, but I'll adjust the help text. Do you want people to be able to set a Max Length of 0?

**Comment by edmistj** [comment]
@dcs619 don't think allowing for 0 is necessary. Thanks for updating the help text.

**Comment by azturner** [comment]
@dcs619 This change is going to cause an issue in the EncryptedTextField type since it is inheriting from the TextFieldType class and adds it's own configuration options (and assumes they are the second/third option). That class should probably be updated so that it does not inherit from TextFieldType and instead inherits from the base FiedlType class. It would then need to add the handling of the IS_PASSWORD option itself rather than getting that functionality from the TextFieldType.  That way you can add the additional option to the TextFieldType.

File: Rock/Field/Types/TextFieldType.cs [hunk 1]
```diff
@@ -74,10 +81,19 @@ public override List<Control> ConfigurationControls()
...
-            { 
-                configurationValues[IS_PASSWORD_KEY].Value = ( (CheckBox)controls[0] ).Checked.ToString();
+            if ( controls != null && controls.Count > 0 )
+            {
+                if ( controls[0] != null && controls[0] is CheckBox )
+                {
+                    configurationValues[IS_PASSWORD_KEY].Value = ( (CheckBox)controls[0] ).Checked.ToString();
+                }
+
+                if ( controls[1] != null && controls[1] is TextBox )
```
Comment (azturner) [inline Rock/Field/Types/TextFieldType.cs]: This should also check to make sure that controls.Count > 1

File: Rock/Field/Types/TextFieldType.cs [hunk 2]
```diff
@@ -111,6 +127,11 @@ public override void SetConfigurationValues( List<Control> controls, Dictionary<
                 {
                     ( (CheckBox)controls[0] ).Checked = configurationValues[IS_PASSWORD_KEY].Value.AsBoolean();
                 }
+
+                if ( controls[1] != null && controls[1] is TextBox && configurationValues.ContainsKey( MAX_LENGTH ) )
```
Comment (azturner) [inline Rock/Field/Types/TextFieldType.cs]: Same thing here

**Comment by dcs619** [comment]
Fixed in 266b9f0. Closing.

File: Rock/Field/Types/TextFieldType.cs (hunk 1 above)
Comment (dcs619) [reply Rock/Field/Types/TextFieldType.cs]: That was already checked at L86

File: Rock/Field/Types/TextFieldType.cs (hunk 2 above)
Comment (dcs619) [reply Rock/Field/Types/TextFieldType.cs]: Checked at L124

**Pull request closed without merging**